nautobot/core/apps/__init__.py

```diff
@@ -131,6 +212,147 @@ def __init__(self, permissions=None):
+                raise TypeError("All items defined in a tab must be an instance of HomePagePanel")
+            self.panels = panels
+
+
+class HomePagePanel(HomePageBase, PermissionsMixin):
+    """Defines properties that can be used for a panel."""
+
+    permissions = []
+    items = []
+
+    @property
+    def initial_dict(self):
+        return {
+            "weight": self.weight,
+            "items": {},
+            "permissions": set(),
+        }
+
+    @property
+    def fixed_fields(self):
+        return ()
+
+    def __init__(self, name, permissions=[], items=None, weight=1000):
+        """Ensure panel properties."""
+        super().__init__(permissions)
+        self.name = name
+        self.weight = weight
+        if items is not None:
+            if not isinstance(items, (list, tuple)):
+                raise TypeError("Items must be passed as a tuple or list.")
+            elif not all(isinstance(item, (HomePageGroup, HomePageItem)) for item in items):
+                raise TypeError("All items defined in a tab must be an instance of HomePageGroup or HomePageItem")
+            self.items = items
+
+
+class HomePageGroup(HomePageBase, PermissionsMixin):
+    """Defines properties that can be used for a panel group."""
+
+    permissions = []
+    items = []
+
+    @property
+    def initial_dict(self):
+        return {
+            "items": {},
+            "permissions": set(),
+            "weight": self.weight,
+        }
+
+    @property
+    def fixed_fields(self):
+        return ()
+
+    def __init__(self, name, permissions=[], items=None, weight=1000):
+        """Ensure group properties."""
+        super().__init__(permissions)
+        self.name = name
+        self.weight = weight
+
+        if items is not None:
+            if not isinstance(items, (list, tuple)):
+                raise TypeError("Items must be passed as a tuple or list.")
+            elif not all(isinstance(item, HomePageItem) for item in items):
+                raise TypeError("All items defined in a tab must be an instance of HomePageItem")
+            self.items = items
+
+
+class HomePageItem(HomePageBase, PermissionsMixin):
+    """Defines properties that can be used for a panel item."""
+
+    permissions = []
+    items = []
+
+    @property
+    def initial_dict(self):
+        return {
+            "custom_code": self.custom_code,
+            "custom_data": self.custom_data,
+            "description": self.description,
+            "link": self.link,
+            "model": self.model,
+            "permissions": self.permissions,
+            "weight": self.weight,
+        }
+
+    @property
+    def fixed_fields(self):
+        return ()
+
+    def __init__(
+        self,
+        name,
+        link=None,
+        model=None,
+        custom_code=None,
+        custom_data=None,
+        description=None,
+        permissions=None,
+        weight=1000,
+    ):
+        """Ensure item properties."""
```

Comment:
Please document arguments in the docstrings for each of these classes - in particular here it would be good to have some explanation and examples of the `custom_code` and `custom_data` keys